### PR TITLE
Partial inlining via function splitting

### PR DESCRIPTION
**Pull request by kripken**
(This is the last "big" optimization on my plate atm for wasm GC.)

This PR helps with functions like this:

```js
function foo(x) {
  if (x) {
    ..
    lots of work here
    ..
  }
}
```

If "lots of work" is large enough, then we won't inline such a
function. However, we may end up calling into the function
only to get a false on that if and immediately exit. So it is useful
to **partially** inline this function, basically by creating a split
of it into a condition part that is inlineable
```js
function foo$inlineable(x) {
  if (x) {
    foo$outlined();
  }
}
```
and an outlined part that is not inlineable:
```js
function foo$outlined(x) {
  ..
  lots of work here
  ..
}
```
We can then inline the inlineable part. That means that a call
like
```js
foo(param);
```
turns into
```js
if (param) {
  foo$outlined();
}
```

In other words, we end up replacing a call and then a check with
a check and then a call. Any time that the condition is false, this
will be a speedup.

The cost here is increased size, as we duplicate the condition
into the callsites. For that reason, only do this when heavily
optimizing for size.

This is a 10% speedup on j2cl. This helps two types of functions
there: Java class inits, which often look like "have I been
initialized before? if not, do all this work", and also assertion
methods which look like "if the input is null, throw an
exception".

### CODE REVIEW

File: src/passes/Inlining.cpp [hunk 1]
```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
...
+    //    ..heavy work..
+    //  }
+    //  B; // an optional final value (which can be a return value)
+    //
+    // where there is a small number of such ifs with arguments A1..A_k, and
+    // A_1..A_k and B (if the final value B exists) are very simple.
+    //
+    // Also, each if body must either be unreachable, or it must have type none
+    // and have no returns. If it is unreachable, for example because it is a
+    // return, then we will just return a value from the outlined function:
```
Comment (tlively) [inline src/passes/Inlining.cpp:616]: IIUC
```suggestion
    // return, then we will just return a value in the inlineable function:
```

File: src/passes/Inlining.cpp [hunk 2]
```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
...
+    // Look for a final item after the ifs.
+    auto* finalItem = getItem(body, numIfs);
+
+    // The final item must be simple (or not exist, which is simple enough).
+    if (finalItem && !isSimple(finalItem)) {
+      return false;
+    }
+
+    // There must be no other items after the optional final one.
+    if (getItem(body, numIfs + (finalItem ? 1 : 0))) {
```
Comment (tlively) [inline src/passes/Inlining.cpp:655]: If there is no final item, then we already know there is nothing after the `If`s:
```suggestion
    if (finalItem && getItem(body, numIfs + 1)) {
```

File: src/passes/Inlining.cpp [hunk 3]
```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
...
+    if (auto* iff = item->dynCast<If>()) {
+      return iff;
+    }
+    return nullptr;
+  }
+
+  // Checks if an expression is very simple - something simple enough that we
+  // are willing to inline it in this optimization. This should basically take
+  // almost no cost at all to compute.
+  bool isSimple(Expression* curr) {
```
Comment (tlively) [inline src/passes/Inlining.cpp:760]: Should the comparison instructions be simple as well?

**Review by tlively — COMMENTED** [review]
I haven't read the tests yet, but here are my comments so far.

File: src/passes/Inlining.cpp [hunk 4]
```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
...
+
+  // Check if an function could be split in order to at least inline part of it,
+  // in a worthwhile manner.
+  //
+  // Even if this returns true, we may not end up inlining the function, as the
+  // main inlining logic has a few other considerations to take into account
+  // (like limitations on which functions can be inlined into in each iteration,
+  // the number of iterations, etc.). Therefore this function will only find out
+  // if we *can* split, but not actually do any splitting.
+  bool canSplit(Function* func, const FunctionInfo& info) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Inlining.cpp]: Good catch!

File: src/passes/Inlining.cpp [hunk 5]
```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
...
+
+    // The default value of split.splittable is false, so if we fail we just
+    // need to return false from this function. If, on the other hand, we can
+    // split, then we will update this split info accordingly.
+    auto& split = splits[func];
+
+    auto* body = func->body;
+
+    // If the body is a block, and we have breaks to that block, then we cannot
+    // outline any code - we can't outline a break without the break's target.
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Inlining.cpp:541]: Not necessarily - one of the patterns below allows returns. The other will check for them.

File: src/passes/Inlining.cpp [hunk 6]
```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
...
+
+    Builder builder(*module);
+
+    // Pattern A: Check if the function begins with
+    //
+    //  if (simple) return;
+    //
+    // TODO: support a return value
+    if (!iff->ifFalse && func->getResults() == Type::none &&
+        iff->ifTrue->is<Return>() && body->is<Block>()) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Inlining.cpp:569]: Good point, we don't need to. In fact lower down we assert on that. I'll move the assert up here and remove the if.

File: src/passes/Inlining.cpp [hunk 7]
```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
...
+      split.splittable = true;
+      // If we were just checking, stop and report success.
+      if (!inlineableOut) {
+        return true;
+      }
+
+      split.inlineable = copyFunction(func, "inlineable-A");
+      auto* outlined = copyFunction(func, "outlined-A");
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Inlining.cpp:577]: `-A` is for "pattern A" as opposed to pattern B that appears below.

Yeah, there is a TODO in the copy function itself for avoiding the extra copying work. It's more work though, and potentially error-prone, so I'm not sure it's worth it. The overhead is wasteful but not that bad.

File: src/passes/Inlining.cpp (hunk 3 above)
Comment (kripken) [reply src/passes/Inlining.cpp:760]: Do you mean the if conditions? This is called on them.

Or what do you mean here?

Comment (tlively) [reply src/passes/Inlining.cpp:763]: I mean things like `i32.lt_u`, `i32.ge_s`, etc.

Comment (kripken) [reply src/passes/Inlining.cpp:763]: Oh, yes, we may want those as well. For now I just added a tiny set of things, but it already handles the main things that j2cl needs. The rest is a TODO (on line 765).

**Review by tlively — APPROVED** [review]

File: test/lit/passes/inlining-optimizing_optimize-level=3.wast [hunk 8]
```diff
@@ -4065,8 +4065,8 @@
  ;; CHECK-NEXT:  (local $21 i32)
  ;; CHECK-NEXT:  (local $22 i32)
  ;; CHECK-NEXT:  (local $23 i32)
- ;; CHECK-NEXT:  (local $24 f64)
- ;; CHECK-NEXT:  (local $25 i32)
+ ;; CHECK-NEXT:  (local $24 i32)
+ ;; CHECK-NEXT:  (local $25 f64)
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/passes/inlining-optimizing_optimize-level=3.wast:4069]: Yeah, sorry about that - for a change like this, inlining more things, the diff on this existing file is quite large.

(Ideally we could somehow make this particular file ignore this change, but otherwise I guess ignoring it in review is second best.)

File: test/lit/passes/inlining_splitting.wast [hunk 9]
```diff
@@ -0,0 +1,1702 @@
...
+  ;; CHECK-NEXT: )
+  (func $call-colliding-name
+    (call $colliding-name (i32.const 0))
+    (call $colliding-name (i32.const 1))
+  )
+
+  ;; CHECK:      (func $colliding-name$byn-outline-A-outlined
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  (func $colliding-name$byn-outline-A-outlined
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/passes/inlining_splitting.wast]: Good catch! I must have renamed in the code and not updated the test...

File: test/lit/passes/inlining_splitting.wast [hunk 10]
```diff
@@ -0,0 +1,1702 @@
...
+  ;; CHECK-NEXT:       )
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (local.get $1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $call-reachable-if-body
+    ;; Note that the above contains
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/passes/inlining_splitting.wast:1032]: Hmm, I think it's nice to show this actually. It is a useful result - if we regress this, we want to know.

File: test/lit/passes/inlining_splitting.wast [hunk 11]
```diff
@@ -0,0 +1,1702 @@
...
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.get $x)
+  ;; CHECK-NEXT: )
+  (func $reachable-if-body-return (param $x anyref) (result anyref)
+    (if
+      (ref.is_null
+        (local.get $x)
+      )
+      (if
+        (i32.const 1)
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/passes/inlining_splitting.wast:1074]: Yes, we don't actually even scan this inner if - we just look for returns inside it.
